### PR TITLE
Let us try to check with the exceptions disabled.

### PR DESCRIPTION
**Pull request by lemire**
Are we properly testing the code when the exceptions are disabled?

### CODE REVIEW

**Comment by lemire** [comment]
Unsurprisingly, our code does not build if we disable exceptions...

**Comment by jkeiser** [comment]
@lemire The tests take care of the flags themselves actually, compiling tests with exceptions on and testing, and then compiling with exceptions off and testing. [You can see the primary tests here,](https://github.com/simdjson/simdjson/blob/master/Makefile#L246-L256) please do take a look for missing tests! There is also a test that exception code fails when exceptions are off, in cmake I believe.

**Comment by lemire** [comment]
Yes, I realize this. What this PR is meant to show is that our tools and tests (e.g., basictests) only compile with exceptions. This is fine, and we can close this PR if we are happy with this result.

**Comment by lemire** [comment]
@jkeiser  I have updated the code so that it now has a chance to pass. I had to do some ugly hacks.

The worse part is this...

```
#if SIMDJSON_EXCEPTIONS
#define ASSERT_EQUAL(ACTUAL, EXPECTED) if ((ACTUAL) != (EXPECTED)) { std::cerr << "Expected " << #ACTUAL << " to be " << (EXPECTED) << ", got " << (ACTUAL) << " instead!" << std::endl; return false; }
#else
#define ASSERT_EQUAL(ACTUAL, EXPECTED) 
#endif
```

(Yes. I disabled the tests for now.)


I think that this is valuable work. We should be able to compile our code and our tests with exceptions disabled. Not just a few toy examples.

**Comment by jkeiser** [comment]
> I think that this is valuable work. We should be able to compile our code and our tests with exceptions disabled. Not just a few toy examples.

I completely agree, we promote the exception-disabled interface so we should test it as much as possible. Should we also consider moving tools to it?

**Comment by lemire** [comment]
Ok. I have re-enabled the tests in basictests.

**Comment by lemire** [comment]
@jkeiser The whole exception-less API for "is<T>" is pretty bad but for now, I just want to test it.

**Comment by jkeiser** [comment]
> @jkeiser The whole exception-less API for "is" is pretty bad but for now, I just want to test it.

Yep.

I just went through and made the tests a bunch more thorough, testing implicit casts, explicit casts, get<>, and exception vs. error: https://github.com/simdjson/simdjson/tree/jkeiser/thorough-type-tests .

I can submit a new PR once this goes through, though.

**Comment by jkeiser** [comment]
@lemire as to the other things, the two big things we should discuss:

1. I don't think we should have both SIMDJSON_DISABLE_EXCEPTIONS and SIMDJSON_EXCEPTIONS in cmake. We should pick one or the other.
2. quickstart.cpp should reflect the docs more closely so that we know the only action we've taken between the two is a copy/paste (minimize potential for error as we update the quickstart).

The fuzzers, I'm fine with your change for now, but we should update them to use no-exception interfaces later. I'll file a bug.

**Comment by lemire** [comment]
> I don't think we should have both SIMDJSON_DISABLE_EXCEPTIONS and SIMDJSON_EXCEPTIONS in cmake. We should pick one or the other.

Ok. Let us go with SIMDJSON_EXCEPTIONS. Commit is in.

**Comment by lemire** [comment]
> quickstart.cpp should reflect the docs more closely so that we know the only action we've taken between the two is a copy/paste (minimize potential for error as we update the quickstart). I think we've both seen way too many projects where the README doesn't even friggin compile, and have been deeply annoyed/offended by it, so we should be extra paranoid in this area.

I did not change the code that appears in the documentation, but this code only works with exceptions enabled. I added similar code without exceptions.

What do you prefer?

1. I can revert it back, but guard it with a check to see if exceptions are enabled.

2. I can keep it as in this commit, but update the quickstart example in our document to include two usage scenarios.

3. I can put it back exactly as you add it, but remove the compilation check in CMake.

I cannot think of another possibility.

**Comment by jkeiser** [comment]
> > quickstart.cpp should reflect the docs more closely so that we know the only action we've taken between the two is a copy/paste (minimize potential for error as we update the quickstart). I think we've both seen way too many projects where the README doesn't even friggin compile, and have been deeply annoyed/offended by it, so we should be extra paranoid in this area.
> 
> I did not change the code that appears in the documentation, but this code only works with exceptions enabled. I added similar code without exceptions.
> 
> What do you prefer?
> 
> 1. I can revert it back, but guard it with a check to see if exceptions are enabled.
> 2. I can keep it as in this commit, but update the quickstart example in our document to include two usage scenarios.
> 3. I can put it back exactly as you add it, but remove the compilation check in CMake.
> 
> I cannot think of another possibility.

The longer way, without changing the primary quick start:

1. Add a no-exception version of quick start somewhere in README or basics.
2. Create a quickstart_noexceptions.cpp
3. Add a test for quickstart_noexceptions.cpp
4. Run the quickstart_noexceptions.cpp test both when exceptions are enabled, and when they are disabled.
5. Run quickstart.cpp compile test only when exceptions are enabled.

The shorter way, changing quick start:

1. Replace the README quickstart with a no-exceptions-only version, removing the exception version. It's not that long, I suppose. Not sure how I feel about it right now, though--the first thing people see is the thing they will try to use first.
2. Replace quickstart.cpp with the same.
3. Compile quickstart.cpp test both with exceptions enabled and disabled.

**Comment by lemire** [comment]
@jkeiser Hmmm... 

Long way it is I suppose.

**Comment by lemire** [comment]
@jkeiser  Long way has been committed. So the README file is untouched. I added the no-exception version in some corner of basics.

File: benchmark/statisticalmodel.cpp [hunk 1]
```diff
@@ -52,7 +52,10 @@ really_inline void simdjson_process_atom(stat_t &s,
   } else if(element.is<double>()) {
     s.float_count++;
   } else if (element.is<bool>()) {
-    if (element.get<bool>()) {
+    simdjson::error_code err;
+    bool v;
+    element.get<bool>().tie(v,err);
```
[the remark this answers is not in the document]
Comment (lemire) [reply benchmark/statisticalmodel.cpp:57]: Yes.

File: fuzz/fuzz_dump.cpp [hunk 2]
```diff
@@ -48,7 +48,7 @@ static void print_json(std::ostream& os, simdjson::dom::element element) {
     }
 }
 extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
-
+#if SIMDJSON_EXCEPTIONS
```
[the remark this answers is not in the document]
Comment (lemire) [reply fuzz/fuzz_dump.cpp:53]: I am not sure I want to touch at the fuzzers in this PR. So I propose that guarding them is probably fine for now.

File: examples/quickstart/quickstart.cpp [hunk 3]
```diff
@@ -1,6 +1,35 @@
...
+#if SIMDJSON_EXCEPTIONS
+
+// version with exceptions:
+
 int main(void) {
   simdjson::dom::parser parser;
   simdjson::dom::element tweets = parser.load("twitter.json");
   std::cout << tweets["search_metadata"]["count"] << " results." << std::endl;
 }
+
```
[the remark this answers is not in the document]
Comment (lemire) [reply examples/quickstart/quickstart.cpp]: > so the noexception version should be in a separate file.

It is (now).

> If we decide we want it, quickstart example with no exceptions should be somewhere in the README or basics.md.

It is (now).

File: CMakeLists.txt [hunk 4]
```diff
@@ -20,6 +20,14 @@ project(simdjson
 #  set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)
 #endif()
 
+### If we are going to pretend to support a mode without exceptions, we should check:
```
[the remark this answers is not in the document]
Comment (lemire) [reply CMakeLists.txt]: > SIMDJSON_EXCEPTIONS already exists in include/CMakeLists.txt. All the global options do. I think I can get behind setting -fno-exceptions as well when this is unset.

Done now.

[the remark this answers is not in the document]
Comment (lemire) [reply CMakeLists.txt]: > I'm also on board if we want to flip it to SIMDJSON_DISABLE_EXCEPTIONS, but if so we should flip the variable we use in include/ as well.

No. I went your way.

File: examples/quickstart/quickstart.cpp [hunk 5]
```diff
@@ -1,6 +1,35 @@
...
+#else
+// version without exceptions:
+
+int main(void) {
+  simdjson::dom::parser parser;
+  simdjson::dom::element tweets;
+  simdjson::error_code error;
+  parser.load("twitter.json").tie(tweets,error);
+  if(error) {
+    std::cerr << "could not load document" << std::endl;
```
[the remark this answers is not in the document]
Comment (lemire) [reply examples/quickstart/quickstart.cpp]: Done.

File: fuzz/fuzz_dump.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkeiser) [reply fuzz/fuzz_dump.cpp:53]: :+1:

**Comment by lemire** [comment]
Let us merge this, we can always improve matters later.

**Comment by jkeiser** [comment]
@lemire it looks like the .circleci config has SIMDJSON_DISABLE_EXCEPTIONS=ON instead of SIMDJSON_EXCEPTIONS=OFF ... and when I flip them, [I get errors](https://app.circleci.com/pipelines/github/simdjson/simdjson/2179/workflows/82ae06fb-a201-4144-b966-b39486154696/jobs/35581). I haven't had time to look and probably won't tonight, but FYI.

**Comment by jkeiser** [comment]
@lemire looks like the issue was from the size() checkin test accidentally used an exception.  Fixed in #720 .

**Pull request merged**